### PR TITLE
Bind Message to ch_message_t and document stuff

### PR DESCRIPTION
**Pull request by ganwell**

### CODE REVIEW

File: libchirp/__init__.py [hunk 1]
```diff
@@ -19,7 +21,7 @@ class Config(object):
     _strings = ('CERT_CHAIN_PEM', 'DH_PARAMS_PEM')
 
     def __init__(self):
-        conf = ffi.new("ch_config_t *")
+        conf = new_nozero("ch_config_t *")
```
Comment (winged) [inline libchirp/__init__.py]: Why are we using this allocator instead of the regular `ffi.new()` one?

File: libchirp/__init__.py [hunk 2]
```diff
@@ -55,3 +57,166 @@ def __getattr__(self, name):
...
+        '_remote_identity',
+    )
+
+    def __init__(self, cmsg=None):
+        self._msg = cmsg
+        self._copy_from_c()
+        if cmsg:
+            lib.ch_msg_free_data(self._msg)
+
+    def _assure_message(self):
```
Comment (winged) [inline libchirp/__init__.py]: Wording: "ensure" would be better fitting for the purpose I think

File: tests/test_message.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    assert msg2.data == data
+    assert msg2.address == ip
+    assert msg2.port == port
+
+
+def test_identity_too_long(message):
+    """test_too_long_identity."""
+    with pytest.raises(AssertionError):
+        message.identity = b'a' * 20
+    message._copy_to_c()
```
Comment (winged) [inline tests/test_message.py]: Is this line really required? Doesn't really do anything anymore.

File: libchirp/__init__.py [hunk 4]
```diff
@@ -55,3 +57,165 @@ def __getattr__(self, name):
...
+        msg.port = self._port
+
+    @property
+    def identity(self):
+        """Get identity."""
+        return self._identity
+
+    @identity.setter
+    def identity(self, value):
+        """Set identity."""
```
Comment (winged) [inline libchirp/__init__.py]: Doc: identity is the node identity, right?

File: libchirp/__init__.py [hunk 5]
```diff
@@ -55,3 +57,165 @@ def __getattr__(self, name):
...
+        self._identity = value
+
+    @property
+    def serial(self):
+        """Get serial."""
+        return self._serial
+
+    @property
+    def header(self):
+        """Get header."""
```
Comment (winged) [inline libchirp/__init__.py]: Doc: is this the parsed format or just a copy of the C struct?

File: libchirp/__init__.py [hunk 6]
```diff
@@ -55,3 +57,165 @@ def __getattr__(self, name):
...
+        return self._address.compressed
+
+    @address.setter
+    def address(self, value):
+        """Set address."""
+        self._address = ip_address(value)
+
+    @property
+    def port(self):
+        """Get port."""
```
Comment (winged) [inline libchirp/__init__.py]: Port of what? remote or ours? (I know, but python-only users might not)

File: libchirp/__init__.py [hunk 7]
```diff
@@ -55,3 +57,165 @@ def __getattr__(self, name):
...
+        return self._port
+
+    @port.setter
+    def port(self, value):
+        """Set port."""
+        assert value >= 0 and value <= 2**16
+        self._port = value
+
+    @property
+    def has_recv_handler(self):
```
Comment (winged) [inline libchirp/__init__.py]: This method could probably be private, no?

**Comment by ganwell** [comment]
@winged I kind of wanted to wait to document everything, till documentation of libchirp is reviewed. And also I don't know how to solve the fact that the doc is more or less duplicated from libchirp. Shall we really copy-paste all this strings? Or reference it? Many problems, I just wanted to wait with the whole documentation.

File: libchirp/__init__.py (hunk 7 above)
Comment (ganwell) [reply libchirp/__init__.py]: No, the user might want to know if the message originated from chirp or not.

File: tests/test_message.py (hunk 3 above)
Comment (ganwell) [reply tests/test_message.py]: Jap. A leftover.

File: libchirp/__init__.py (hunk 1 above)
Comment (ganwell) [reply libchirp/__init__.py]: I added a comment up where the allocator is created.

Comment (winged) [reply libchirp/__init__.py]: Okay, makes sense then

File: libchirp/__init__.py (hunk 7 above)
Comment (winged) [reply libchirp/__init__.py]: Then I'd rename it accordingly. Users from python-land may not know (or want to know) what a recv handler is or does

**Comment by ganwell** [comment]
@winged I am working on the doc.

About has_recv_handler, the doc is

True if the message has a chirp-recv-buffer attached and therefore you have to call self.release(). It also means this message is created by chirp. By default chirp will release the message automatically when the handler-callback returns. See self.release().

Possible names

* needs_release
* has_recv_buffer
* is_received_message
* created_by_chirp

class Config() will get a python only setting AUTO_RELEASE. If the user sets this to False. He has to use self.release().

**Comment by ganwell** [comment]
@winged I just realized that the messages identity should be read-only (while documenting) so documenting early might be a good thing after all.

**Comment by ganwell** [comment]
@winged Can you please review

* Document class Config
* Document class Message

I will resolve the whole has_recv_handler/release issue when libchirp is updated see concretecloud/chirp#257

File: libchirp/__init__.py [hunk 8]
```diff
@@ -16,7 +16,114 @@
...
+       conn_size = conn_buffers_size + config.MAX_MSG_SIZE
+
+       With the default config and SSL conn_buffers_size should be about
+       64k + 16k + 2k + 32k -> 114k. Derived from documentation, no
+       measurement done.
+
+    .. py:attribute:: BIND_V6
+
+       Override IPv6 bind address. String representation expected, parsed by
+       :py:class:`ipaddress.ip_address`.
```
[the remark this answers is not in the document]
Comment (ganwell) [reply libchirp/__init__.py]: I don't think so. The exceptions and limitations of this are defined by ip_address. It important to be able to find out want is happening on a problem.

File: libchirp/__init__.py [hunk 9]
```diff
@@ -16,7 +16,114 @@
...
+
+    .. py:attribute:: BIND_V4
+
+       Override IPv4 bind address. String representation expected, parsed by
+       :py:class:`ipaddress.ip_address`.
+
+    .. py:attribute:: IDENTITY
+
+       Override the chirp-nodes IDENTITY (this chirp instance). By default
+       all chars are 0, which means chirp will generate a IDENTITY. Python
```
[the remark this answers is not in the document]
Comment (ganwell) [reply libchirp/__init__.py]: Its not even true. by default an identity is created.

File: libchirp/__init__.py [hunk 10]
```diff
@@ -16,7 +16,114 @@
...
+
+       IO related timeout: Sending messages, connecting to remotes. (float)
+
+    .. py:attribute:: PORT
+
+       Port for listening to connections. (uint16_t)
+
+    .. py:attribute:: BACKLOG
+
+       TCP-listen socket backlog. (uint8_t)
```
[the remark this answers is not in the document]
Comment (ganwell) [reply libchirp/__init__.py]: Documented below

File: libchirp/__init__.py [hunk 11]
```diff
@@ -16,7 +16,114 @@
...
+       ACKNOWLEDGE=1. (uint8_t)
+
+    .. py:attribute:: ACKNOWLEDGE
+
+       Acknowledge messages. Default True. Makes chirp connection-synchronous.
+       See :ref:`modes-of-operation`. Python boolean expected.
+
+    .. py:attribute:: DISABLE_SIGNALS
+
+       By default chirp closes on SIGINT (Ctrl-C) and SIGTERM. Python boolean
```
[the remark this answers is not in the document]
Comment (ganwell) [reply libchirp/__init__.py]: Documented below the default is False

File: libchirp/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (winged) [reply libchirp/__init__.py]: Disagree, but OK.

[the remark this answers is not in the document]
Comment (ganwell) [reply libchirp/__init__.py]: But if you really think I can remove it.

**Pull request merged**